### ISSUE TITLE
Unresolved type error for a class in transitive dependency

### ISSUE THREAD

**Issue by lwasyl** [google/dagger#3262]
Dagger 2.45

Repro project: https://github.com/lwasyl/dagger-transitive-dependencies/tree/2634c3b3fb316fc877fc0e0919e79e36946efc31

If you run `./gradlew presentation:assemble`, the build will fail with a message that it can't resolve `ApiModel`. If you change Dagger version to 2.40.5 in `settings.gradle`, the same command will succeed.

I see in the changelog more strict validation for unresolved annotations, but nothing for types. In this repro, there are no scopes or qualifiers, so I don't know if the change is expected or not

**Comment by bcorso** [comment]
Hi @lwasyl,

Sorry, you're right that we didn't mention this specific change in the release notes. This was more of an incidental change related to the scopes and qualifier work.

In general, we're starting to lean more heavily towards validating all Dagger-related types (i.e. types within elements annotated with `@Inject`, `@Provides`, `@Binds`, etc.). In this case, the `ApiModel` shows up as the return type of an `@Provides` method, which is why it is being validated now:

```kotlin
@Module
abstract class CoreModule {
    companion object {
        @Provides internal fun coreInterface() = object : InternalCoreInterface<ApiModel> {}
    }
}
```

Although in this case we may not need `ApiModel` to be resolvable for correctness, we'd like to still require that the type be resolvable since it is within an `@Provides` method that Dagger is processing and we may need it to be resolvable for future validation.

For now, you can add `ApiModel` either as a direct `implementation` dependency of `:app`, or as an `api` dependency of `:core`. Feel free to also give feedback on how much of an impact this has on your particular app (e.g. how often you run into it).

**Comment by lwasyl** [comment]
Hey @bcorso, sorry for late response.

> In general, we're starting to lean more heavily towards validating all Dagger-related types

That means that a module where a component is generated will require compile-time access to all types that are part of that `@Component` (all of its `@Module`s)?

> Feel free to also give feedback on how much of an impact this has on your particular app (e.g. how often you run into it).

This is surprisingly common for us at least with API models specifically. What we do (and would _like_ to keep doing) is keep `core -> networking` dependency an `implementation` to avoid leaking out the api models outside of the modules that should use them. Granted, the models _leaked_ via Dagger, but they aren't part of the public API of the user code itself (in the example repo, the `InternalCoreInterface` is `internal`, `InternalCoreInterface<ApiModel>` field/parameter is private, and even the `CoreInterfaceImpl` is `internal`. Dagger-generated classes were only used by Dagger, so that looked fine. Now, we'd have to give up that safety by explicitly giving access to `ApiModel`s to many modules that shouldn't use it.

Can you elaborate on the _safety_ that these new checks provide? Release notes mention component/qualifier-related bugs which I understand, are there similar bugs related to unresolvable types like in my example too?

Overall I can see why the stricter validation is reasonable - after all, our `ApiModel` _is_ part of the module api 😕 I'd like to reconcile that somehow with the safety of hiding `ApiModel`, but maybe the only _correct_ way of doing that is generating an additional component (although that'd be a pretty big change)

**Comment by GianfrancoMS** [comment]
@bcorso Is it possible that you can revert this change/behavior or put it behind a flag? It's causing a lot of problems, specially for a big multi-module project.

**Comment by bcorso** [comment]
> Can you elaborate on the safety that these new checks provide? Release notes mention component/qualifier-related bugs which I understand, are there similar bugs related to unresolvable types like in my example too?

@lwasyl, one example is `@Binds` methods, where our processors need to validate that the parameter type is assignable to the return type at compile time. In particular, in order to even perform that validation we need to first ensure that both the parameter type and the return type are fully resolvable.

> Is it possible that you can revert this change/behavior or put it behind a flag? It's causing a lot of problems, specially for a big multi-module project.

If this is causing a lot of issues, I think we're willing to relax this constraint specifically for the case that @lwasyl mentioned for the return type of an `@Provides` method, since we don't currently rely on that type being fully resolvable for any of our processing (although, as mentioned above, we'll still need for other cases like `@Binds`).

@GianfrancoMS would this also solve your issues, or are you running into other cases besides the `@Provides` return types?

**Comment by GianfrancoMS** [comment]
@bcorso I think it's the same problem. Here are some logs after updating to 2.41

```
> Task :chat:kaptDebugKotlin FAILED
error: ComponentProcessingStep was unable to process 'com.rappi.chat.di.LegacyChatFeatureComponent' because 'com.rappi.uxcam.api.UxCamObfuscator' could not be resolved.
  
  Dependency trace:
      => element (INTERFACE): com.rappi.base.di.BaseComponent
      => element (METHOD): uxCamObfuscator()
      => type (EXECUTABLE method): ()com.rappi.uxcam.api.UxCamObfuscator
      => type (ERROR return type): com.rappi.uxcam.api.UxCamObfuscator
  
  If type 'com.rappi.uxcam.api.UxCamObfuscator' is a generated type, check above for compilation errors that may have prevented the type from being generated. Otherwise, ensure that type 'com.rappi.uxcam.api.UxCamObfuscator' is on your classpath.
```

**Comment by bcorso** [comment]
@GianfrancoMS it looks like that issue is slightly different since the method is not an `@Provides` method.

Can you describe what `BaseComponent` is? 

I'm guessing it's either an interface that your `@Component`-annotated class/interface extends or its a component dependency you pass to `@Component.dependencies`?

**Comment by GianfrancoMS** [comment]
@bcorso `BaseComponent` is a base component that feature legacy components extend. 

```
interface BaseComponent {

    fun analyticsLogger(): AnalyticsAction

    fun analyticsConfig(): AnalyticsConfig

    fun mainFunnel(): MainFunnel

    fun uxCamObfuscator(): UxCamObfuscator

    // Other dependencies.
}
```

```
@FeatureScope
@Component(
    modules = [
        FactoryModule::class,
        AndroidSupportInjectionModule::class,
        ChatActivityBuilder::class,
        OrdersModule::class,
        FirebaseDatabaseModule::class
    ],
    dependencies = [LegacyChatFeatureComponent.Parent::class]
)
interface LegacyChatFeatureComponent {

    fun androidInjector(): DispatchingAndroidInjector<Any>

    interface Parent : BaseComponent {
        fun getHomeOrdersController(): HomeOrdersController
        fun getOrderTrackingController(): OrderTrackingController
        fun realtimeConfigurationController(): RealtimeConfigurationController
    }

    @Component.Factory
    interface Factory {
        fun create(parent: Parent): LegacyChatFeatureComponent
    }
}
```

**Comment by lwasyl** [comment]
> one example is `@Binds` methods, where our processors need to validate that the parameter type is assignable to the return type at compile time. In particular, in order to even perform that validation we need to first ensure that both the parameter type and the return type are fully resolvable.

Just fyi, for us relaxing this requirement for `@Provides` but _not_ for `@Binds` methods will not change much. We have lots of `@Binds` methods too, which means we'd probably expose the _offending_ types regardless. Consider that it might be good to avoid surprises when migrating between `@Provides` and `@Binds` (I would be pretty surprised to see compilation errors after "optimizing" a `@Provides` into a `@Binds`)

I understand all the reasoning too, I'll just say that I kind of hoped that the validation for `@Binds` could happen within the module where the binding method is processed — after all, that's the place where we fully expect all the classes to be resolvable in compile time, because that's where we write _our_ code. I'm not sure how feasible it is or how it would affect rest of Dagger, but it _feels_ like a reasonable approach — I find it difficult to figure out a scenario in which such validation would be correct at the module level but invalid at the component level

**Comment by bcorso** [comment]
[@lwasyl](https://github.com/lwasyl), [@GianfrancoMS](https://github.com/GianfrancoMS) just wanted to give an update.

# TL;DR

Unfortunately, I think the proposed changes may be a bit more problematic than I initially thought.

In particular, even if we relax Dagger’s validation, javac ends up failing conditionally depending on how exactly we implement the generated code (see details below). This makes our generated code more fragile to changes and also means that toggling on/off different build flags, e.g. [fastInit mode](https://dagger.dev/dev-guide/compiler-options#fastinit-mode), may cause unexpected breakages. Thus, while this may have worked okay for you in the past, it’s fragile and there’s no guarantee that some future change to our generated code won't accidentally break you.

I do list some workarounds on our side that could likely avoid these issues (see below), but given that it adds complexity to our code base, reduces the readability of our generated code, and likely  requires a lot more time/effort we’re going to continue to validate the types for these case.

# Javac Issues

As an example of how javac can conditionally fail based on our generated code, I'll use `CoreInterface` from  the [sample app](https://github.com/lwasyl/dagger-transitive-dependencies/blob/2634c3b3fb316fc877fc0e0919e79e36946efc31/core/src/main/java/com/example/CoreInterface.kt) you linked.

Before looking at each scenario, it's important to understand that Dagger generates a factory class for each binding that contains both a static `create()` method and a static `newInstance()` method that our generated code uses to create either an instance of the factory (e.g. `Foo_Factory`) or type (e.g. `Foo`) respectively.

## Scenario 1

In some cases we satisfy the request for a binding using `Factory#newInstance()`, e.g.

```java
@Override
public CoreInterface getCoreInterface() {
  return CoreInterfaceImpl_Factory.newInstance(
      CoreModule_internalCoreInterfaceFactory.newInstance());
}
```
 
This scenario fails javac because the call to `CoreModule_internalCoreInterfaceFactory.newInstance()` returns `InternalCoreInterface<ApiModel>` which references `ApiModel` (which is not on the classpath). Note that this happens even though we don't explicitly reference `ApiModel` anywhere in the generated code. Javac will fail with an error like:

```
error: cannot access ApiModel
    return CoreInterfaceImpl_Factory.newInstance(
       CoreModule_internalCoreInterfaceFactory.newInstance());

  class file for ApiModel not found
```
 
## Scenario 2

In other cases, we need to use a `Provider` to satisfy the request using `Factory.create()`, e.g.

```java
private Provider<CoreInterface> coreInterfaceProvider;

void initialize() {
  coreInterfaceProvider =
      CoreInterfaceImpl_Factory.create(CoreModule_internalCoreInterfaceFactory.create());
}

@Override
public CoreInterface getCoreInterface() {
  return coreInterfaceProvider.get();
}
```
 
In this case javac does not fail. It turns out that we get a bit lucky because the `CoreModule_internalCoreInterfaceFactory.create()` returns the subclass type `CoreModule_internalCoreInterfaceFactory` rather than its supertype, `Provider<InternalCoreInterface<ApiModel>>`, which would have caused javac to fail due to the reference to `ApiModel`.


# Workarounds

One workaround to this issue is that we could change `Foo_Factory#newInstance()` to always return `Object` rather than `Foo` to avoid javac from having to know about the type when compiling the generated component. Unfortunately, we would have to do this for every binding since we can't know apriori which classes will not be on the classpath. This also means that all of the input parameters for `#newInstance()` and `#create()` would also have to be `Object`, which would add a lot more casts to the generated code. We would also need some version skew logic to check whether the generated factory was from a new enough version of Dagger to include these changes.

**Comment by bcorso** [comment]
@freszu you are hitting https://github.com/google/dagger/issues/970.

By default, we validate all of your transitive components to check for things like invalid scope cycles. However, there's a flag to disable that validation. Add this to module `C` to disable it:

```groovy
kapt {
    arguments {
        arg("dagger.validateTransitiveComponentDependencies", "DISABLED")
    }
}
```
